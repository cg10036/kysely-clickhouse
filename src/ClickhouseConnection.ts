import {
  CompiledQuery,
  DatabaseConnection,
  QueryResult
} from 'kysely';

import { createClient } from '@clickhouse/client'
import { ClickhouseDialectConfig } from '.';
import { NodeClickHouseClient } from '@clickhouse/client/dist/client';


export class ClickhouseConnection implements DatabaseConnection {
  #client: NodeClickHouseClient

  constructor(config: ClickhouseDialectConfig) {
    this.#client = createClient(config.options)
  }

  prepareQuery<O>(compiledQuery: CompiledQuery): string {
    let i = 0
    const compiledSql = compiledQuery.sql.replace(/\?/g, () => {
      const param = compiledQuery.parameters[i++]

      if (typeof param === 'number') {
        return `${param}`
      }

      // should never happen
      if (typeof param !== 'string') {
        return `'${JSON.stringify(param)}'`
      }

      return `'${param.replace(/'/gm, `\\'`).replace(/\\"/g, '\\\\"')}'`
    })

    return compiledSql
  }

  async executeQuery<O>(compiledQuery: CompiledQuery): Promise<QueryResult<O>> {
    const query = this.prepareQuery(compiledQuery)

    const resultSet = await this.#client.query({
      query,
      format: 'JSONEachRow',
    })
    const rows: O[] = await resultSet.json()

    return {
      rows
    }

  }

  async beginTransaction() {
    throw new Error('Transactions are not supported.');
  }

  async commitTransaction() {
    throw new Error('Transactions are not supported.');
  }

  async rollbackTransaction() {
    throw new Error('Transactions are not supported.');
  }

  async *streamQuery<O>(compiledQuery: CompiledQuery, chunkSize: number): AsyncIterableIterator<QueryResult<O>> {
    const query = this.prepareQuery(compiledQuery)

    const resultSet = await this.#client.query({
      query,
      format: 'JSONEachRow',
    })

    const stream = resultSet.stream()

    for await (const row of stream) {
      yield {
        rows: [row as O],
      }
    }
  }
}