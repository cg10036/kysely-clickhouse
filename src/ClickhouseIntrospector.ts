import {
  MysqlIntrospector
} from 'kysely'

export class ClickhouseIntrospector extends MysqlIntrospector {}
