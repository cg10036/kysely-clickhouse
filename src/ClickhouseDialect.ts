import {
  DatabaseIntrospector,
  Dialect,
  Driver,
  Kysely,
  MysqlAdapter,
  MysqlQueryCompiler,
  QueryCompiler
} from 'kysely';

import { ClickhouseDriver } from './ClickhouseDriver';
import { ClickhouseIntrospector } from './ClickhouseIntrospector';
import type { NodeClickHouseClientConfigOptions } from '@clickhouse/client/dist/config';

export interface ClickhouseDialectConfig {
  options?: NodeClickHouseClientConfigOptions;
}

export class ClickhouseDialect implements Dialect {
  #config: ClickhouseDialectConfig;

  constructor(config?: ClickhouseDialectConfig) {
    this.#config = config ?? {};
  }

  createAdapter() {
    return new MysqlAdapter();
  }

  createDriver(): Driver {
    return new ClickhouseDriver(this.#config);
  }

  createQueryCompiler(): QueryCompiler {
    return new MysqlQueryCompiler();
  }

  createIntrospector(db: Kysely<any>): DatabaseIntrospector {
    return new ClickhouseIntrospector(db, this.#config);
  }
}



