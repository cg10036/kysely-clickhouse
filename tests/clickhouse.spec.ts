import { expect, test } from 'vitest'

import {  ClickhouseDialect } from '../src'
import { Kysely } from 'kysely';

const kysely = new Kysely<any>({
  dialect: new ClickhouseDialect({
    options: {
      url: 'http://localhost:8123',
      username: 'founderpath',
      password: 'founderpath',
      database: 'founderpath',
    }
  })
})

test('introspection', async () => {
  const tables = await kysely.introspection.getTables()

  expect(tables).toMatchSnapshot()
})