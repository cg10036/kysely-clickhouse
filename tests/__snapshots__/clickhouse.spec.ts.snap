// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`introspection 1`] = `
[
  {
    "columns": [
      {
        "comment": undefined,
        "dataType": "String",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": false,
        "name": "name",
      },
      {
        "comment": undefined,
        "dataType": "String",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": false,
        "name": "timestamp",
      },
    ],
    "isView": false,
    "name": "_migrations",
    "schema": "founderpath",
  },
  {
    "columns": [
      {
        "comment": undefined,
        "dataType": "String",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": false,
        "name": "id",
      },
      {
        "comment": undefined,
        "dataType": "Int32",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": false,
        "name": "is_locked",
      },
    ],
    "isView": false,
    "name": "_migrations_lock",
    "schema": "founderpath",
  },
  {
    "columns": [
      {
        "comment": undefined,
        "dataType": "String",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": false,
        "name": "id",
      },
      {
        "comment": undefined,
        "dataType": "String",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": false,
        "name": "bank_account_id",
      },
      {
        "comment": undefined,
        "dataType": "String",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": false,
        "name": "remote_id",
      },
      {
        "comment": undefined,
        "dataType": "String",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": false,
        "name": "fpath_account_id",
      },
      {
        "comment": undefined,
        "dataType": "Date",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": false,
        "name": "date",
      },
      {
        "comment": undefined,
        "dataType": "String",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": false,
        "name": "original_description",
      },
      {
        "comment": undefined,
        "dataType": "UInt8",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": false,
        "name": "pending",
      },
      {
        "comment": undefined,
        "dataType": "String",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": false,
        "name": "processing_status",
      },
      {
        "comment": undefined,
        "dataType": "Int64",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": false,
        "name": "amount",
      },
      {
        "comment": undefined,
        "dataType": "String",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": false,
        "name": "iso_currency_code",
      },
      {
        "comment": undefined,
        "dataType": "String",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": false,
        "name": "unofficial_currency_code",
      },
      {
        "comment": undefined,
        "dataType": "String",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": false,
        "name": "bank_account_category_id",
      },
      {
        "comment": undefined,
        "dataType": "Date",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": false,
        "name": "authorized_date",
      },
      {
        "comment": undefined,
        "dataType": "String",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": false,
        "name": "name",
      },
      {
        "comment": undefined,
        "dataType": "String",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": false,
        "name": "merchant_name",
      },
      {
        "comment": undefined,
        "dataType": "String",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": false,
        "name": "pending_transaction_id",
      },
      {
        "comment": undefined,
        "dataType": "String",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": false,
        "name": "transaction_type",
      },
      {
        "comment": undefined,
        "dataType": "String",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": false,
        "name": "account_owner",
      },
      {
        "comment": undefined,
        "dataType": "Int64",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": false,
        "name": "amount_usd",
      },
      {
        "comment": undefined,
        "dataType": "String",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": false,
        "name": "transaction_category",
      },
      {
        "comment": undefined,
        "dataType": "String",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": false,
        "name": "transaction_status",
      },
      {
        "comment": undefined,
        "dataType": "DateTime",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": false,
        "name": "created_at",
      },
      {
        "comment": undefined,
        "dataType": "DateTime",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": false,
        "name": "updated_at",
      },
      {
        "comment": undefined,
        "dataType": "String",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": false,
        "name": "ntropy_batch_id",
      },
      {
        "comment": undefined,
        "dataType": "DateTime",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": false,
        "name": "enriched_at",
      },
      {
        "comment": undefined,
        "dataType": "String",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": false,
        "name": "enriched_by",
      },
      {
        "comment": undefined,
        "dataType": "String",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": false,
        "name": "enrichment_cat1",
      },
      {
        "comment": undefined,
        "dataType": "String",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": false,
        "name": "enrichment_cat2",
      },
      {
        "comment": undefined,
        "dataType": "String",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": false,
        "name": "enrichment_cat3",
      },
      {
        "comment": undefined,
        "dataType": "String",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": false,
        "name": "enrichment_cat4",
      },
      {
        "comment": undefined,
        "dataType": "String",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": false,
        "name": "enrichment_merchant_name",
      },
    ],
    "isView": false,
    "name": "bank_account_transactions",
    "schema": "founderpath",
  },
  {
    "columns": [
      {
        "comment": undefined,
        "dataType": "UInt64",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": false,
        "name": "company_id",
      },
      {
        "comment": undefined,
        "dataType": "Date",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": false,
        "name": "date",
      },
      {
        "comment": undefined,
        "dataType": "DateTime",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": false,
        "name": "updated_at",
      },
      {
        "comment": undefined,
        "dataType": "String",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": false,
        "name": "customer_email",
      },
      {
        "comment": undefined,
        "dataType": "String",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": false,
        "name": "remote_customer_id",
      },
      {
        "comment": "Source of customer, e.g. Stripe, Chargebee",
        "dataType": "String",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": false,
        "name": "source",
      },
      {
        "comment": "Status of customer within month based on change in MRR from previous period. [new_customer,stable,expansion,contraction,churn]",
        "dataType": "String",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": false,
        "name": "customer_status",
      },
      {
        "comment": undefined,
        "dataType": "Decimal",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": false,
        "name": "mrr",
      },
      {
        "comment": undefined,
        "dataType": "Decimal",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": false,
        "name": "ltv",
      },
      {
        "comment": "Amount of unchanged MRR from previous period",
        "dataType": "Decimal",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": false,
        "name": "mrr_stable",
      },
      {
        "comment": "Amount of new MRR from previous period",
        "dataType": "Decimal",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": false,
        "name": "mrr_new",
      },
      {
        "comment": "Amount of churned MRR from previous period",
        "dataType": "Decimal",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": false,
        "name": "mrr_churned",
      },
      {
        "comment": "Amount of expanded MRR from previous period",
        "dataType": "Decimal",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": false,
        "name": "mrr_expanded",
      },
      {
        "comment": "Amount of contracted MRR from previous period",
        "dataType": "Decimal",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": false,
        "name": "mrr_contracted",
      },
    ],
    "isView": false,
    "name": "customer_metrics",
    "schema": "founderpath",
  },
  {
    "columns": [
      {
        "comment": undefined,
        "dataType": "UInt64",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": false,
        "name": "company_id",
      },
      {
        "comment": undefined,
        "dataType": "String",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": false,
        "name": "customer_name",
      },
      {
        "comment": undefined,
        "dataType": "String",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": false,
        "name": "customer_email",
      },
      {
        "comment": undefined,
        "dataType": "String",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": false,
        "name": "remote_customer_id",
      },
      {
        "comment": undefined,
        "dataType": "Decimal",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": false,
        "name": "mrr",
      },
      {
        "comment": undefined,
        "dataType": "Decimal",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": false,
        "name": "ltv",
      },
      {
        "comment": undefined,
        "dataType": "DateTime",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": false,
        "name": "started_at",
      },
      {
        "comment": undefined,
        "dataType": "DateTime",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "ended_at",
      },
      {
        "comment": undefined,
        "dataType": "DateTime",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "cancelled_at",
      },
      {
        "comment": undefined,
        "dataType": "String",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": false,
        "name": "plan_remote_id",
      },
      {
        "comment": undefined,
        "dataType": "String",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": false,
        "name": "plan_name",
      },
      {
        "comment": undefined,
        "dataType": "String",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": false,
        "name": "plan_term",
      },
      {
        "comment": undefined,
        "dataType": "Int32",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": false,
        "name": "seats",
      },
      {
        "comment": undefined,
        "dataType": "String",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": false,
        "name": "source",
      },
      {
        "comment": undefined,
        "dataType": "String",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": false,
        "name": "status",
      },
      {
        "comment": undefined,
        "dataType": "String",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": false,
        "name": "domain",
      },
      {
        "comment": undefined,
        "dataType": "String",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": false,
        "name": "country",
      },
      {
        "comment": "Is the domain valid",
        "dataType": "UInt8",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": false,
        "name": "is_valid",
      },
      {
        "comment": "Has the domain been crawled",
        "dataType": "UInt8",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": false,
        "name": "is_crawled",
      },
      {
        "comment": "Is the domain parked, any domain name that is not linked to any unique website or web service",
        "dataType": "UInt8",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": false,
        "name": "is_parked",
      },
      {
        "comment": "Is the domain from an easy cms service, e.g. wix or squarespace",
        "dataType": "UInt8",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": false,
        "name": "is_easy_cms",
      },
      {
        "comment": "Is the domain from a freemail, ex. gmail",
        "dataType": "UInt8",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": false,
        "name": "is_freemail",
      },
      {
        "comment": undefined,
        "dataType": "DateTime",
        "hasDefaultValue": true,
        "isAutoIncrementing": false,
        "isNullable": false,
        "name": "updated_at",
      },
    ],
    "isView": false,
    "name": "customers",
    "schema": "founderpath",
  },
]
`;
